### ISSUE TITLE
Shader portability (and performance portability).

### ISSUE THREAD

**Issue by krogovin** [gpuweb/gpuweb#895]
I am filing this issue for the purpose of opening the can of worms on shader and shader-performance portability. One theme that has come up is the expectation of portability in that incorrect use of the API would not work on one platform but work on another, the example issue being #862 . For compute shaders, there is the issue of size of shared memory (and for that matter how much padding arrays of structs in shared memory will take). Another issue which is more common is array access.

Consider the following code

```
uniform int M, MM;
uniform isampler1D S, T;
out float V;
float my_array[SomeMaxSize];
int size = 0;

my_array[0] = some_ugly_computation(0, ..);
for (int i = 1; i < M; ++i)
{
    my_array[i] = some_ugly_computation(my_array[i - 1], ..);
}

float R = 0.0;
for (int i =0; i < MM; ++i)
 {
   R += texture(S, my_array[texture(T, i, 0)], 0);
}
```

this is perfectly legal shader code as far as I can tell and will run and work on Intel GPU's and drivers.

However, in the past, random array access from shaders (especially fragment) of local arrays was dicy on lots of GPU's.

Another ugly issue is how much shared memory is available to compute? Will it be queryable and a compute application adapts to the value? Or is the value the minimum across all minimums of the backend API's? Same question for work group sizes as well.

What is the plan for these kind of things?

**Comment by kvark** [comment]
These are great concerns, thank you for explaining them clearly. I believe this is going to be the most challenging aspect of portability for us to achieve.

> However, in the past, random array access from shaders (especially fragment) of local arrays was dicy on lots of GPU's.

Can you tell more? Which GPU families/platforms are affected, and what happens there?

> Another ugly issue is how much shared memory is available to compute?

We haven't decided yet, but one likely outcome is that being just one of the limits, with a reasonable minimum established by the spec. Investigation is welcome!

**Comment by Kangz** [comment]
> Another ugly issue is how much shared memory is available to compute? Will it be queryable and a compute application adapts to the value? Or is the value the minimum across all minimums of the backend API's? Same question for work group sizes as well.

To piggy back on @kvark's answer, these will be in `GPULimits`, queriable from the `GPUAdapter` and applications will be able to opt into higher limits. Vulkan has similar limits: `maxComputeSharedMemorySize`, `maxComputeWorkGroupCount`, `maxComputeWorkGroupInvocations`, and `maxComputeWorkGroupSize`. D3D12 and Metal have these limits per features level / GPU family IIRC.

If a GPU / driver doesn't support random access in local arrays, then it is something browsers will have to find a workaround for (or disable WebGPU on that combination). The CTS should definitely have tests that exercise this type of fun shaders.

**Comment by krogovin** [comment]
The main issue for *local* array, not uniform arrays, is that a shader compiler will try to make local variables live in the register file. In the dark days of GL2 and DirectX9, dynamic indexing of local arrays almost always failed: either the shader compiler complained "internal compiler error/too complicated shader" or the resulting shader rendered utterly incorrect. To that end a number of bizarre tricks where used like 

```
array = vec4(value0, value1, value2, value3);

for (int i = 0; i < N; ++i)
{
   v = array.x;
   something_with(v, ...);
   array.xyz = array.yzw;
}
``` 

which I hated to do.

What I can see is that in order so support dynamically indexed *local* arrays (or for that matter shared memory in a compute shader) is that the memory model requirement is going to be quite high to support that use case since, i.e. the analogue of pointer support.

FWIW, the way the Intel GPU and drivers handle dynamic indexed local arrays is to allocate scratch space (essentially the size being SizeOfScratchPerShaderInvocation * MaxNumberOfShadersInvocationsAlive) and issue memory reads and writes to that scratch space; all dynamic arrays end up living on that scratch space.

**Comment by Kangz** [comment]
> The main issue for local array, not uniform arrays

Yes, that doesn't change the problem. If a GPU / driver has an issue with it, the browser needs to workaround it, or it needs to blocklist the GPU / driver combination. See also https://github.com/gpuweb/gpuweb/issues/863#issuecomment-651230099

**Comment by grorg** [comment]
Briefly discussed at the [2020-08-25 meeting](https://docs.google.com/document/d/13lWCWtooVE1kT6TGta16gM-dDXjRdUKt9KG_9cTcTkY/edit?usp=sharing). Resolution was that there needs to be investigation before we can address this completely. In the meantime we'll try to maintain performance portability as much as possible.

**Comment by Kangz** [comment]
Tentatively closing this: limits have been added for the max compute workgroup size and max workgroup storage size. The issues with indexable vs. non-indexable registers is annoying, but WGSL doesn't have this restriction and it is WGSL / driver compilers that will have to handle it. (Tint already does to make FXC happy)
